File: Algorithms/Implementations/cavity_map.c
#include <stdio.h>

void modify_map(int length, char cavity_map[length][length])
{
    int current;
    for (int m = 1; m < length-1; m ++)
    {
        for (int n = 1; n < length-1; n++)
        {
            current = cavity_map[m][n] - '0';
            if (cavity_map[m-1][n] - '0' < current && cavity_map[m][n-1] - '0'< current && cavity_map[m][n+1] - '0'< current && cavity_map[m+1][n] - '0' < current)
            {
                cavity_map[m][n] = 'X';
            }
        }
    }

}

void print_map(int length, char cavity_map[length][length])
{
    for (int i = 0; i < length; i++)
    {
        for (int j = 0; j < length; j++)
        {
            printf("%c", cavity_map[i][j]);
        }
        printf("\n");
    }
}

int main()
{
    int length;

    scanf("%d", &length);
    char cavity_map[length][length];
    for (int i = 0; i < length; i++)
    {
        scanf("%s", cavity_map[i]);
    }

    modify_map(length, cavity_map);
    print_map(length, cavity_map);

    return 0;
}

File: Algorithms/Implementations/cavity_map.py
def modify_map(cavity_map):
    modified_map = []
    size_of_map = len(cavity_map)
    for i in xrange(size_of_map):
        if i == 0 or i == size_of_map - 1:
            modified_map.append(cavity_map[i])
        else: 
            row = []
            for j in xrange(size_of_map):
                if j == 0 or j == size_of_map - 1:
                    row.append(cavity_map[i][j])
                else:
                    # check the numbers around it
                    center = cavity_map[i][j]
                    all_small = True
                    surrounding = [cavity_map[i-1][j],
                                   cavity_map[i][j-1],
                                   cavity_map[i][j+1],
                                   cavity_map[i+1][j]]

                    for number in surrounding:
                        if number >= center:
                            all_small = False

                    if all_small:
                        row.append("X")
                    else:
                        row.append(center)
            modified_map.append(row)
    return modified_map 
if __name__ == "__main__":
    length = int(raw_input().strip())
    cavity_map = []
    for i in xrange(length):
        cavity_map.append(map(int, list(raw_input().strip())))
    new_map = modify_map(cavity_map)
    for line in new_map:
        print "".join(map(str, line))
